identity-aware-proxy/IAPClient/IAPClient.cs

```diff
@@ -17,49 +17,45 @@
 using System.Threading.Tasks;
 
-namespace GoogleCloudSamples
+public class IAPClient
 {
-    class IAPClient
+    /// <summary>
+    /// Obtains an OIDC token for authenticating and IAP request.
+    /// </summary>
+    /// <param name="iapClientId">The client id observed on 
+    /// https://console.cloud.google.com/apis/credentials. </param>
+    /// <param name="credentialsFilePath">Path to the credentials .json file
+    /// downloaded from https://console.cloud.google.com/apis/credentials.
+    /// </param>
+    /// <param name="uri">HTTP uri to fetch.</param>
+    /// <param name="cancellationToken">The token to propagate operation cancel notifications.</param>
+    /// <returns>The http response message.</returns>
+    public async Task<HttpResponseMessage> InvokeRequestAsync(
+        string iapClientId, string credentialsFilePath, string uri, CancellationToken cancellationToken = default)
     {
-        /// <summary>
-        /// Authenticates using the client id and credentials, then fetches
-        /// the uri.
-        /// </summary>
-        /// <param name="iapClientId">The client id observed on 
-        /// https://console.cloud.google.com/apis/credentials. </param>
-        /// <param name="credentialsFilePath">Path to the credentials .json file
-        /// downloaded from https://console.cloud.google.com/apis/credentials.
-        /// </param>
-        /// <param name="uri">HTTP uri to fetch.</param>
-        /// <returns>The http response body as a string.</returns>
-        public async Task<string> InvokeRequestAsync(string iapClientId, string credentialsFilePath, string uri)
-        {
-            // Read credentials from the credentials .json file.
-            ServiceAccountCredential saCredential;
-            using (var fs = new FileStream(credentialsFilePath, FileMode.Open, FileAccess.Read))
-            {
-                saCredential = ServiceAccountCredential.FromServiceAccountData(fs);
-            }
-
-            // Request an OIDC token for the Cloud IAP-secured client ID.
-            OidcToken oidcToken = await saCredential.GetOidcTokenAsync(OidcTokenOptions.FromTargetAudience(iapClientId)).ConfigureAwait(false);
-            // Always request the string token from the OIDC token, the OIDC token will refresh the string token if it expires.
-            string token = await oidcToken.GetAccessTokenAsync().ConfigureAwait(false);
-
-            // Include the OIDC token in an Authorization: Bearer header to 
-            // IAP-secured resource
-            using (var httpClient = new HttpClient())
-            {
-                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
-                string response = await httpClient.GetStringAsync(uri).ConfigureAwait(false);
-                return response;
-            }
-        }
+        // Read credentials from the credentials .json file.
+        GoogleCredential credential = await GoogleCredential

```

Comment:
Would it be worth separating the "one-time-setup" part from the "on every request" part? I'm guessing that `credential` and `oidcToken` only need to be initialized once, but everything later would be per request.

Reply:
Yep, leaving a comment also.